### PR TITLE
RUMM-2808 add network tracing sampling rate

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Adds an option to set a tracing sampling rate on the `ddUrlTransfer` component, allowing having a lower number of traces and controlling the costs of DD

### CODE REVIEW

**Review by estherk15 — APPROVED** [review]

File: library/source/wrapper/DdUrlTransfer.bs [hunk 1]
```diff
@@ -18,13 +18,25 @@ class DdUrlTransfer
     ' @param datadogRumAgent (object) the Datadog RumAgent node to report
     ' requests to
     ' ----------------------------------------------------------------
-    sub new(datadogRumAgent as object)
+    sub new(datadogRumAgent as object, tracingSamplingRate = 100.0 as double)
         m.roUrlTransfer = CreateObject("roUrlTransfer")
         m.datadogRumAgent = datadogRumAgent
+        m.tracingSamplingRate = tracingSamplingRate
+        ddLogWarning("tracingSamplingRate:" + tracingSamplingRate.toStr())
+        ddLogWarning("m.tracingSamplingRate:" + m.tracingSamplingRate.toStr())
```
Comment (0xnm) [inline library/source/wrapper/DdUrlTransfer.bs:26]: probably should be removed?

File: library/source/wrapper/DdUrlTransfer.bs [hunk 2]
```diff
@@ -18,13 +18,25 @@ class DdUrlTransfer
...
         ' TODO define trace sampling rate
         ' TODO define first party hosts
     end sub
 
+    ' ----------------------------------------------------------------
+    ' Sets the URL to use for the transfer request.
+    '
+    ' @param url (string) The URL to be used for the transfer request
+    ' ----------------------------------------------------------------
```
Comment (0xnm) [inline library/source/wrapper/DdUrlTransfer.bs:35]: description to be updated

File: library/source/wrapper/DdUrlTransfer.bs [hunk 3]
```diff
@@ -624,6 +616,27 @@ class DdUrlTransfer
     sub ClearCookies()
         m.roUrlTransfer.ClearCookies()
     end sub
+
+    sub _traceRequest()
+        m.traceId = generateUniqueId()
+        m.spanId = generateUniqueId()
+        rndTrace = (Rnd(101) - 1) ' Rnd(n) returns a number between 1 and n (both inclusive)
+        ddLogWarning("rndTrace:" + rndTrace.toStr())
+        ddLogWarning("m.tracingSamplingRate:" + m.tracingSamplingRate.toStr())
```
Comment (0xnm) [inline library/source/wrapper/DdUrlTransfer.bs:625]: should it be kept? the same for warning entries below.

File: library/source/wrapper/DdUrlTransfer.bs [hunk 4]
```diff
@@ -624,6 +616,27 @@ class DdUrlTransfer
...
+            ddLogWarning("Network request trace for " + m.GetUrl() + " is KEPT ")
+            m.roUrlTransfer.AddHeader("x-datadog-trace-id", m.traceId)
+            m.roUrlTransfer.AddHeader("x-datadog-parent-id", m.spanId)
+            m.roUrlTransfer.AddHeader("x-datadog-sampling-priority", "1")
+            m.roUrlTransfer.AddHeader("x-datadog-origin", "rum")
+        else
+            ddLogWarning("Network request trace for " + m.GetUrl() + " is DISCARDED  ")
+            m.traceId = invalid
+            m.spanId = invalid
+            m.spanId = generateUniqueId()
```
Comment (0xnm) [inline library/source/wrapper/DdUrlTransfer.bs:636]: why do we still generate span ID after setting it to `invalid`?

**Review by 0xnm — COMMENTED** [review]
lgtm, I've added some comments.

File: dist/source/wrapper/DdUrlTransfer.brs [hunk 5]
```diff
@@ -16,12 +16,23 @@ function __DdUrlTransfer_builder()
     ' @param datadogRumAgent (object) the Datadog RumAgent node to report
     ' requests to
     ' ----------------------------------------------------------------
-    instance.new = sub(datadogRumAgent as object)
+    instance.new = sub(datadogRumAgent as object, tracingSamplingRate = 100.0 as double)
         m.roUrlTransfer = CreateObject("roUrlTransfer")
         m.datadogRumAgent = datadogRumAgent
+        m.tracingSamplingRate = tracingSamplingRate
+        ddLogWarning("tracingSamplingRate:" + tracingSamplingRate.toStr())
+        ddLogWarning("m.tracingSamplingRate:" + m.tracingSamplingRate.toStr())
```
Comment (plousada) [inline dist/source/wrapper/DdUrlTransfer.brs:24]:
```suggestion

```

File: dist/source/wrapper/DdUrlTransfer.brs [hunk 6]
```diff
@@ -579,11 +570,31 @@ function __DdUrlTransfer_builder()
     instance.ClearCookies = sub()
         m.roUrlTransfer.ClearCookies()
     end sub
+    instance._traceRequest = sub()
+        m.traceId = generateUniqueId()
+        m.spanId = generateUniqueId()
+        rndTrace = (Rnd(101) - 1) ' Rnd(n) returns a number between 1 and n (both inclusive)
+        ddLogWarning("rndTrace:" + rndTrace.toStr())
+        ddLogWarning("m.tracingSamplingRate:" + m.tracingSamplingRate.toStr())
+        if (rndTrace < m.tracingSamplingRate)
```
Comment (plousada) [inline dist/source/wrapper/DdUrlTransfer.brs:579]: We currently don't have any firstPartyHost mechanism to help decide which requests to trace in Roku?

File: dist/source/wrapper/DdUrlTransfer.brs [hunk 7]
```diff
@@ -579,11 +570,31 @@ function __DdUrlTransfer_builder()
     instance.ClearCookies = sub()
         m.roUrlTransfer.ClearCookies()
     end sub
+    instance._traceRequest = sub()
+        m.traceId = generateUniqueId()
+        m.spanId = generateUniqueId()
+        rndTrace = (Rnd(101) - 1) ' Rnd(n) returns a number between 1 and n (both inclusive)
+        ddLogWarning("rndTrace:" + rndTrace.toStr())
+        ddLogWarning("m.tracingSamplingRate:" + m.tracingSamplingRate.toStr())
```
Comment (plousada) [inline dist/source/wrapper/DdUrlTransfer.brs:578]:
```suggestion
```

**Review by plousada — APPROVED** [review]

File: library/source/wrapper/DdUrlTransfer.bs (hunk 1 above)
Comment (xgouchet) [reply library/source/wrapper/DdUrlTransfer.bs:26]: true

File: library/source/wrapper/DdUrlTransfer.bs (hunk 2 above)
Comment (xgouchet) [reply library/source/wrapper/DdUrlTransfer.bs:35]: Nice catch

File: library/source/wrapper/DdUrlTransfer.bs (hunk 4 above)
Comment (xgouchet) [reply library/source/wrapper/DdUrlTransfer.bs:636]: That's a typo, nice catch

**Review by 0xnm — APPROVED** [review]

**Pull request merged**